libr/util/axml.c

```diff
@@ -0,0 +1,281 @@
+			char *value = resource_value (pool, data, &a.value);
+			r_strbuf_appendf (sb, "%s=%s", key, value);
+			if (i != count - 1) {
+				r_strbuf_appendf (sb, " ");
+			}
+			free (key);
+			free (value);
+		}
+	} else {
+		r_strbuf_appendf (sb, "</%s", name);
+	}
+
+	r_strbuf_appendf (sb, ">\n");
+	free (name);
+	return true;
+}
+
+R_API char *r_axml_decode(const ut8 *data, const ut64 data_size) {
+	string_pool_t *pool = NULL;
+	RStrBuf *sb = NULL;
+
+	if (!data || !data_size) {
+		eprintf ("Invalid buffer or size\n");
+		return NULL;
+	}
+
+	RBuffer *buffer = r_buf_new_with_pointers (data, data_size, false);
+	if (buffer == NULL) {
+		eprintf ("Error allocating RBuffer\n");
+		goto error;
+	}
+
+	ut64 offset = 0;
+
+	chunk_header_t header = {0};
+	if (r_buf_fread_at (buffer, offset, (ut8 *)&header, "ssi", 1) != sizeof(header)) {
+		goto bad;
+	}
+
+	if (header.type != TYPE_XML) {
+		goto bad;
+	}
+
+	offset += sizeof(header);
+
+	sb = r_strbuf_new ("");
+
+	while (offset < data_size) {
+		if (r_buf_fread_at (buffer, offset, (ut8 *)&header, "ssi", 1) != sizeof(header)) {
+			goto bad;
+		}
+
+		ut16 type = header.type;
+
+		// After reading the original chunk header, read the type-specific
+		// header
+		offset += sizeof(header);
+
+		switch (type) {
+		case TYPE_STRING_POOL: {
+			ut16 header_size = header.size;
+			if (header_size == 0 || header_size > data_size) {
+				goto bad;
+			}
+
+			pool = malloc (header_size);
+			if (pool == NULL) {
+				goto bad;
+			}
+
+			if (r_buf_read_at (buffer, offset, (void *)pool, header_size) != header_size) {
+				goto bad;
+			}
+		} break;
+		case TYPE_START_ELEMENT: {
+			// The string pool must be the first header
+			if (pool == NULL) {
+				goto bad;
+			}
+
+			ut16 header_size = header.size;
+			if (header_size == 0 || header_size > data_size) {
+				goto bad;
+			}
+
+			start_element_t *element = malloc (header_size);
+			if (element == NULL) {
+				goto bad;
+			}
+
+			if (r_buf_read_at (buffer, offset, (void *)element, header_size) != header_size) {
+				free(element);
+				goto bad;
+			}
+
+			if (!dump_element (sb, pool, data, element, true)) {
+				free(element);
+				goto bad;
+			}
+
+			depth++;
+
+			free (element);
+		} break;
+		case TYPE_END_ELEMENT: {
+			depth--;
+
+			end_element_t end;
+			if (r_buf_read_at (buffer, offset, (void *)&end, sizeof(end)) != sizeof(end)) {
+				goto bad;
+			}
+
+			// The beginning of the start and end element structs
+			// are the same, so we can use this interchangably
+			if (!dump_element (sb, pool, data, (start_element_t *)&end, false)) {
+				goto bad;
+			}
+		} break;
+		default:
+			eprintf ("Type is not recognized: %#x\n", type);
+		}
+
+		offset += header.size - sizeof(header);
+	}
+
+	r_buf_free (buffer);
+	if (pool != NULL)
+		free (pool);
+	return r_strbuf_drain (sb);
+bad:
+	eprintf("Invalid Android Binary XML\n");
+error:
+	if (buffer != NULL)
+		r_buf_free (buffer);
+	if (pool != NULL)
+		free (pool);
+	if (sb != NULL)
```

Comment:
```suggestion
```

Reply:
No, if `sb` is NULL then `drain` will dereference a NULL pointer